File: src/core/Shaders/random.fs
// -----------------------------------------------------------------------------
// Uniform declarations and constants
// -----------------------------------------------------------------------------
uniform int samplingTechnique; // 0: PCG, 1: Halton, 2: Sobol, 3: Blue Noise
uniform sampler2D blueNoiseTexture;
uniform ivec2 blueNoiseTextureSize;

// Golden ratio constants for dimension decorrelation
const float PHI = 1.61803398875;
const float INV_PHI = 0.61803398875;
const float INV_PHI2 = 0.38196601125;

// Sobol sequence direction vectors
const uint V[ 32 ] = uint[ 32 ]( 2147483648u, 1073741824u, 536870912u, 268435456u, 134217728u, 67108864u, 33554432u, 16777216u, 8388608u, 4194304u, 2097152u, 1048576u, 524288u, 262144u, 131072u, 65536u, 32768u, 16384u, 8192u, 4096u, 2048u, 1024u, 512u, 256u, 128u, 64u, 32u, 16u, 8u, 4u, 2u, 1u );

// Primes for hashing (carefully chosen to avoid correlations)
const uint PRIME1 = 2654435761u;
const uint PRIME2 = 3266489917u;
const uint PRIME3 = 668265263u;
const uint PRIME4 = 374761393u;

// -----------------------------------------------------------------------------
// Basic random number generation
// -----------------------------------------------------------------------------

// PCG (Permuted Congruential Generator) hash function
uint pcg_hash( uint state ) {
    state = state * 747796405u + 2891336453u;
    state = ( ( state >> ( ( state >> 28u ) + 4u ) ) ^ state ) * 277803737u;
    state = ( state >> 22u ) ^ state;
    return state;
}

// Wang hash for additional mixing
uint wang_hash( uint seed ) {
    seed = ( seed ^ 61u ) ^ ( seed >> 16u );
    seed *= 9u;
    seed = seed ^ ( seed >> 4u );
    seed *= 0x27d4eb2du;
    seed = seed ^ ( seed >> 15u );
    return seed;
}

// Generate random float between 0 and 1
float RandomValue( inout uint state ) {
    state = pcg_hash( state );
    return float( state >> 8 ) * ( 1.0 / 16777216.0 );
}

// Generate random float with better precision
float RandomValueHighPrecision( inout uint state ) {
    uint s1 = pcg_hash( state );
    state = s1;
    uint s2 = pcg_hash( state );
    // Combine two 24-bit values for 48-bit precision
    return ( float( s1 >> 8 ) + float( s2 >> 8 ) * ( 1.0 / 16777216.0 ) ) * ( 1.0 / 16777216.0 );
}

// -----------------------------------------------------------------------------
// Directional sampling functions
// -----------------------------------------------------------------------------

// Generate random point in unit circle
vec2 RandomPointInCircle( inout uint rngState ) {
    float angle = RandomValue( rngState ) * TWO_PI;
    vec2 pointOnCircle = vec2( cos( angle ), sin( angle ) );
    return pointOnCircle * sqrt( RandomValue( rngState ) );
}

// -----------------------------------------------------------------------------
// Blue noise sampling with proper multi-dimensional support
// -----------------------------------------------------------------------------

// Cranley-Patterson rotation for decorrelation
vec2 cranleyPatterson2D( vec2 p, vec2 offset ) {
    return fract( p + offset );
}

// Improved blue noise sampling that properly uses all parameters
vec4 sampleBlueNoiseRaw( vec2 pixelCoords, int sampleIndex, int bounceIndex ) {
    // Create dimension-specific offsets using golden ratio
    vec2 dimensionOffset = vec2( fract( float( sampleIndex ) * INV_PHI ), fract( float( bounceIndex ) * INV_PHI2 ) );

    // Frame-based decorrelation with better hash
    uint frameHash = wang_hash( pcg_hash( uint( frame ) ) );
    vec2 frameOffset = vec2( float( frameHash & 0xFFFFu ) / 65536.0, float( ( frameHash >> 16 ) & 0xFFFFu ) / 65536.0 );

    // Scale offsets to texture size
    vec2 scaledDimOffset = dimensionOffset * vec2( blueNoiseTextureSize );
    vec2 scaledFrameOffset = frameOffset * vec2( blueNoiseTextureSize );

    // Combine all offsets with proper toroidal wrapping
    vec2 coords = mod( pixelCoords + scaledDimOffset + scaledFrameOffset, vec2( blueNoiseTextureSize ) );

    // Ensure positive coordinates and fetch
    ivec2 texCoord = ivec2( floor( coords ) );
    return texelFetch( blueNoiseTexture, texCoord, 0 );
}

// Get a single float value from blue noise (for 1D sampling)
float sampleBlueNoise1D( vec2 pixelCoords, int sampleIndex, int dimension ) {
    vec4 noise = sampleBlueNoiseRaw( pixelCoords, sampleIndex, dimension / 4 );
    int component = dimension % 4;
    return component == 0 ? noise.x : component == 1 ? noise.y : component == 2 ? noise.z : noise.w;
}

// Get 2D blue noise sample with dimension offset
vec2 sampleBlueNoise2D( vec2 pixelCoords, int sampleIndex, int dimensionBase ) {
    // For 2D sampling, we need to carefully select components to maintain blue noise properties
    vec4 noise = sampleBlueNoiseRaw( pixelCoords, sampleIndex, dimensionBase / 2 );

    // Use different component pairs based on dimension
    int pairIndex = ( dimensionBase / 2 ) % 6;
    switch( pairIndex ) {
        case 0:
            return noise.xy;
        case 1:
            return noise.zw;
        case 2:
            return noise.xz;
        case 3:
            return noise.yw;
        case 4:
            return noise.xw;
        case 5:
            return noise.yz;
    }
    return noise.xy; // fallback
}

// Progressive blue noise sampling for temporal accumulation
vec2 sampleProgressiveBlueNoise( vec2 pixelCoords, int currentSample, int maxSamples ) {
    // Determine which "slice" of the blue noise we're in
    float progress = float( currentSample ) / float( max( 1, maxSamples ) );
    int temporalSlice = int( progress * 16.0 ); // 16 temporal slices

    // Use different regions of blue noise for different sample counts
    vec2 sliceOffset = vec2( float( temporalSlice % 4 ) * 0.25, float( temporalSlice / 4 ) * 0.25 );

    // Scale to texture space and add pixel-specific offset
    vec2 scaledOffset = sliceOffset * vec2( blueNoiseTextureSize );
    vec2 coords = mod( pixelCoords + scaledOffset, vec2( blueNoiseTextureSize ) );

    vec4 noise = sampleBlueNoiseRaw( coords, currentSample, 0 );

    // Apply additional Cranley-Patterson rotation for better distribution
    uint seed = pcg_hash( uint( currentSample ) ^ wang_hash( uint( maxSamples ) ) );
    vec2 rotation = vec2( float( seed & 0xFFFFu ) / 65536.0, float( ( seed >> 16 ) & 0xFFFFu ) / 65536.0 );

    return cranleyPatterson2D( noise.xy, rotation );
}

// -----------------------------------------------------------------------------
// Low-discrepancy sequence generators
// -----------------------------------------------------------------------------

// Halton sequence generator with Owen scrambling
float haltonScrambled( int index, int base, uint scramble ) {
    float result = 0.0;
    float f = 1.0;
    int i = index;

    while( i > 0 ) {
        f /= float( base );
        // Apply digit scrambling
        int digit = i % base;
        digit = int( wang_hash( uint( digit ) ^ scramble ) % uint( base ) );
        result += f * float( digit );
        i = int( floor( float( i ) / float( base ) ) );
    }
    return result;
}

// Owen scrambling for Sobol sequence
uint owen_scramble( uint x, uint seed ) {
    x = x ^ ( x * 0x3d20adeau );
    x += seed;
    x *= ( seed >> 16 ) | 1u;
    x ^= x >> 15;
    x *= 0x5851f42du;
    x ^= x >> 12;
    x *= 0x4c957f2du;
    x ^= x >> 18;
    return x;
}

// Owen-scrambled Sobol sequence
float owen_scrambled_sobol( uint index, uint dimension, uint seed ) {
    uint result = 0u;
    for( int i = 0; i < 32; ++ i ) {
        if( ( index & ( 1u << i ) ) != 0u ) {
            result ^= V[ i ] << dimension;
        }
    }
    result = owen_scramble( result, seed );
    return float( result ) / 4294967296.0;
}

vec2 owen_scrambled_sobol2D( uint index, uint seed ) {
    return vec2( owen_scrambled_sobol( index, 0u, seed ), owen_scrambled_sobol( index, 1u, seed ) );
}

// -----------------------------------------------------------------------------
// Multi-dimensional sampling interface
// -----------------------------------------------------------------------------

// Get N-dimensional sample (up to 4D)
vec4 getRandomSampleND( vec2 pixelCoord, int sampleIndex, int bounceIndex, inout uint rngState, int dimensions, int preferredTechnique ) {
    int technique = ( preferredTechnique != - 1 ) ? preferredTechnique : samplingTechnique;
    vec4 result = vec4( 0.0 );

    switch( technique ) {
        case 0: // PCG
            for( int i = 0; i < dimensions; i ++ ) {
                result[ i ] = RandomValue( rngState );
            }
            break;

        case 1: // Halton
        {
            uint scramble = pcg_hash( uint( pixelCoord.x ) + uint( pixelCoord.y ) * uint( resolution.x ) );
            int primes[ 4 ] = int[ 4 ]( 2, 3, 5, 7 );
            for( int i = 0; i < dimensions; i ++ ) {
                result[ i ] = haltonScrambled( sampleIndex, primes[ i ], scramble );
            }
        }
        break;

        case 2: // Sobol
        {
            uint seed = pcg_hash( uint( pixelCoord.x ) + uint( pixelCoord.y ) * uint( resolution.x ) );
            for( int i = 0; i < dimensions; i ++ ) {
                result[ i ] = owen_scrambled_sobol( uint( sampleIndex ), uint( i ), seed );
            }
        }
        break;

        case 3: // Blue Noise
        {
                // For blue noise, we need to carefully handle dimensions
            int dimensionOffset = bounceIndex * 4; // Each bounce uses up to 4 dimensions

            if( dimensions <= 2 ) {
                vec2 _sample = sampleBlueNoise2D( pixelCoord, sampleIndex, dimensionOffset );
                result.xy = _sample;
            } else {
                    // For 3D/4D, fetch two 2D samples
                vec2 _sample1 = sampleBlueNoise2D( pixelCoord, sampleIndex, dimensionOffset );
                vec2 _sample2 = sampleBlueNoise2D( pixelCoord, sampleIndex, dimensionOffset + 2 );
                result = vec4( _sample1, _sample2 );
            }
        }
        break;
    }

    return result;
}

// -----------------------------------------------------------------------------
// Hybrid sampling methods
// -----------------------------------------------------------------------------

// Combine quasi-random and pseudo-random sampling with blue noise awareness
vec2 HybridRandomSample2D( inout uint state, int sampleIndex, int pixelIndex ) {
    vec2 quasi;

    if( samplingTechnique == 2 ) { // Sobol
        uint seed = pcg_hash( uint( pixelIndex ) );
        quasi = owen_scrambled_sobol2D( uint( sampleIndex ), seed );
    } else if( samplingTechnique == 1 ) { // Halton
        uint scramble = wang_hash( uint( pixelIndex ) );
        quasi = vec2( haltonScrambled( sampleIndex, 2, scramble ), haltonScrambled( sampleIndex, 3, scramble ) );
    } else if( samplingTechnique == 3 ) { // Blue noise
        vec2 pixelCoord = vec2( float( pixelIndex % int( resolution.x ) ), float( pixelIndex / int( resolution.x ) ) );
        return sampleBlueNoise2D( pixelCoord, sampleIndex, 0 );
    } else { // PCG fallback
        return vec2( RandomValue( state ), RandomValue( state ) );
    }

    // Add small random offset for better convergence
    vec2 pseudo = vec2( RandomValue( state ), RandomValue( state ) );
    return fract( quasi + pseudo * 0.01 ); // Small perturbation
}

// -----------------------------------------------------------------------------
// Main sampling interface functions
// -----------------------------------------------------------------------------

// Get random sample based on preferred technique (2D)
vec2 getRandomSample( vec2 pixelCoord, int sampleIndex, int bounceIndex, inout uint rngState, int preferredTechnique ) {
    vec4 sample4D = getRandomSampleND( pixelCoord, sampleIndex, bounceIndex, rngState, 2, preferredTechnique );
    return sample4D.xy;
}

// Get stratified sample with proper blue noise support
vec2 getStratifiedSample( vec2 pixelCoord, int rayIndex, int totalRays, inout uint rngState ) {
    if( totalRays == 1 ) {
        return getRandomSample( pixelCoord, rayIndex, 0, rngState, - 1 );
    }

    // Calculate strata dimensions
    int strataX = int( sqrt( float( totalRays ) ) );
    int strataY = ( totalRays + strataX - 1 ) / strataX;

    int strataIdx = rayIndex % ( strataX * strataY );
    int sx = strataIdx % strataX;
    int sy = strataIdx / strataX;

    // Base stratified position
    vec2 strataPos = vec2( float( sx ), float( sy ) ) / vec2( float( strataX ), float( strataY ) );

    // Get jitter based on sampling technique
    vec2 jitter;
    if( samplingTechnique == 3 ) { // Blue noise
        // Use progressive blue noise for stratified sampling
        jitter = sampleProgressiveBlueNoise( pixelCoord, rayIndex, totalRays );
    } else {
        jitter = getRandomSample( pixelCoord, rayIndex, 0, rngState, - 1 );
    }

    jitter /= vec2( float( strataX ), float( strataY ) );

    return strataPos + jitter;
}

// Get decorrelated seed with better mixing
uint getDecorrelatedSeed( vec2 pixelCoord, int rayIndex, uint frame ) {
    // Use multiple primes for better decorrelation
    uint pixelSeed = uint( pixelCoord.x ) * PRIME1 + uint( pixelCoord.y ) * PRIME2;
    uint raySeed = uint( rayIndex ) * PRIME3;
    uint frameSeed = frame * PRIME4;

    // Multiple rounds of hashing for better quality
    uint seed = wang_hash( pixelSeed );
    seed = pcg_hash( seed ^ raySeed );
    seed = wang_hash( seed + frameSeed );

    return seed;
}

// -----------------------------------------------------------------------------
// Specialized sampling functions
// -----------------------------------------------------------------------------

// Get sample optimized for primary rays (pixel anti-aliasing)
vec2 getPrimaryRaySample( vec2 pixelCoord, int sampleIndex, int totalSamples, inout uint rngState ) {
    if( samplingTechnique == 3 ) { // Blue noise
        return sampleProgressiveBlueNoise( pixelCoord, sampleIndex, totalSamples );
    }
    return getStratifiedSample( pixelCoord, sampleIndex, totalSamples, rngState );
}

// Get sample optimized for BRDF sampling
vec2 getBRDFSample( vec2 pixelCoord, int sampleIndex, int bounceIndex, inout uint rngState ) {
    // BRDF sampling benefits from different dimensions than pixel sampling
    int dimensionOffset = 2 + bounceIndex * 2; // Start at dimension 2

    if( samplingTechnique == 3 ) { // Blue noise
        return sampleBlueNoise2D( pixelCoord, sampleIndex, dimensionOffset );
    }
    return getRandomSample( pixelCoord, sampleIndex, bounceIndex, rngState, - 1 );
}

// Get sample for Russian roulette (1D)
float getRussianRouletteSample( vec2 pixelCoord, int sampleIndex, int bounceIndex, inout uint rngState ) {
    if( samplingTechnique == 3 ) { // Blue noise
        // Use a different dimension for Russian roulette
        return sampleBlueNoise1D( pixelCoord, sampleIndex, bounceIndex * 4 + 3 );
    }
    return RandomValue( rngState );
}

File: rayzee/src/TSL/Random.js
// Three.js Transpiler r182

import { uniform, texture, float, If, wgslFn, uint, TWO_PI, cos, sin, vec2, sqrt, fract, mod, ivec2, select, int, vec4, mix } from 'three/tsl';
import { DataTexture, FloatType } from 'three';

// -----------------------------------------------------------------------------
// Uniform declarations and constants
// -----------------------------------------------------------------------------

export const samplingTechniqueUniform = uniform( 0, 'int' );
const samplingTechnique = samplingTechniqueUniform;

// 0: PCG, 1: Halton, 2: Sobol, 3: Blue Noise

// 1x1 placeholder — real texture assigned later via .value = ...
const _placeholderData = new Float32Array( [ 0.5, 0.5, 0.5, 1.0 ] );

const _placeholderScalar = new DataTexture( _placeholderData, 1, 1 );
_placeholderScalar.type = FloatType;
_placeholderScalar.needsUpdate = true;

const _placeholderVec2 = new DataTexture( new Float32Array( [ 0.5, 0.5, 0.0, 1.0 ] ), 1, 1 );
_placeholderVec2.type = FloatType;
_placeholderVec2.needsUpdate = true;

// STBN (Spatiotemporal Blue Noise) atlas textures — Heitz 2019
// Each atlas: 1024×1024, 8×8 grid of 128×128 tiles, 64 temporal slices
// Scalar atlas: single-channel (R) — optimal for 1D decisions (RR, lobe selection)
// Vec2 atlas: two-channel (R,G) — decorrelated 2D pairs (direction sampling Xi)
export const stbnScalarTextureNode = texture( _placeholderScalar );
stbnScalarTextureNode.setUpdateMatrix( false );

export const stbnVec2TextureNode = texture( _placeholderVec2 );
stbnVec2TextureNode.setUpdateMatrix( false );

// R2 quasi-random sequence constants (Roberts 2018) — optimal 2D additive offsets
const R2_A1 = float( 0.7548776662466927 );
const R2_A2 = float( 0.5698402909980532 );

// Sobol sequence direction vectors using function lookup for compatibility

// Sobol direction vectors for first dimension — exact port of GLSL
export const getSobolDirectionVector = /*@__PURE__*/ wgslFn( `
	fn getSobolDirectionVector( index: i32 ) -> u32 {

		switch ( index ) {
			case 0:  { return 2147483648u; }
			case 1:  { return 1073741824u; }
			case 2:  { return  536870912u; }
			case 3:  { return  268435456u; }
			case 4:  { return  134217728u; }
			case 5:  { return   67108864u; }
			case 6:  { return   33554432u; }
			case 7:  { return   16777216u; }
			case 8:  { return    8388608u; }
			case 9:  { return    4194304u; }
			case 10: { return    2097152u; }
			case 11: { return    1048576u; }
			case 12: { return     524288u; }
			case 13: { return     262144u; }
			case 14: { return     131072u; }
			case 15: { return      65536u; }
			case 16: { return      32768u; }
			case 17: { return      16384u; }
			case 18: { return       8192u; }
			case 19: { return       4096u; }
			case 20: { return       2048u; }
			case 21: { return       1024u; }
			case 22: { return        512u; }
			case 23: { return        256u; }
			case 24: { return        128u; }
			case 25: { return         64u; }
			case 26: { return         32u; }
			case 27: { return         16u; }
			case 28: { return          8u; }
			case 29: { return          4u; }
			case 30: { return          2u; }
			default: { return          1u; }
		}

	}
` );

// -----------------------------------------------------------------------------
// Basic random number generation
// -----------------------------------------------------------------------------
// PCG (Permuted Congruential Generator) hash function

export const pcgHash = /*@__PURE__*/ wgslFn( `
	fn pcgHash( state: u32 ) -> u32 {

		var s = state;
		s = s * 747796405u + 2891336453u;
		s = ( ( s >> ( ( s >> 28u ) + 4u ) ) ^ s ) * 277803737u;
		s = ( s >> 22u ) ^ s;
		return s;

	}
` );

// Wang hash for additional mixing

export const wang_hash = /*@__PURE__*/ wgslFn( `
	fn wang_hash( seed: u32 ) -> u32 {

		var s = seed;
		s = ( s ^ 61u ) ^ ( s >> 16u );
		s = s * 9u;
		s = s ^ ( s >> 4u );
		s = s * 0x27d4eb2du;
		s = s ^ ( s >> 15u );
		return s;

	}
` );

// OPTIMIZED: Fast random value for hot paths - uses simpler hash for performance
// Performance gain: ~40% faster than full PCG for non-critical samples

// Plain JS functions (not Fn()) — they inline at the call site where `state` is a
// mutable .toVar() variable. This replicates GLSL `inout uint state` semantics:
// the caller's rngState advances on every call.

export const RandomValueFast = ( state ) => {

	// Simple multiply-with-carry generator - much faster than PCG
	state.assign( state.mul( 1664525 ).add( 1013904223 ) );
	return float( state.shiftRight( 8 ) ).mul( 1.0 / 16777216.0 );

};

// Generate random float between 0 and 1 with full PCG quality

export const RandomValue = ( state ) => {

	state.assign( pcgHash( { state } ) );
	return float( state.shiftRight( 8 ) ).mul( 1.0 / 16777216.0 );

};

// Generate random float with better precision

export const RandomValueHighPrecision = ( state ) => {

	// Capture s1 immediately (.toVar()) before advancing state, so it isn't
	// re-evaluated lazily after the second pcgHash advances state further.
	const s1 = pcgHash( { state } ).toVar();
	state.assign( s1 );
	const s2 = pcgHash( { state } ).toVar();

	// Combine two 24-bit values for 48-bit precision
	return float( s1.shiftRight( 8 ) ).add( float( s2.shiftRight( 8 ) ).mul( 1.0 / 16777216.0 ) ).mul( 1.0 / 16777216.0 );

};

// -----------------------------------------------------------------------------
// Directional sampling functions
// -----------------------------------------------------------------------------
// OPTIMIZED: Fast random point in unit circle using simpler RNG for hot paths

export const RandomPointInCircle = ( rngState ) => {

	// Use fast RNG for circle sampling - adequate quality for DOF/sampling
	// .toVar() captures angle immediately so cos/sin don't read state after the 2nd advance
	const angle = RandomValueFast( rngState ).mul( TWO_PI ).toVar();
	const pointOnCircle = vec2( cos( angle ), sin( angle ) );

	return pointOnCircle.mul( sqrt( RandomValueFast( rngState ) ) );

};

// -----------------------------------------------------------------------------
// STBN atlas sampling — proper spatiotemporal blue noise
// -----------------------------------------------------------------------------
// Atlas layout: 8×8 grid of 128×128 tiles = 1024×1024 texture.
// Temporal axis: frame % 64 selects tile (true STBN temporal decorrelation).
// Spatial decorrelation: R2 quasi-random offset keyed on dimension + sample index.

const computeSTBNAtlasCoord = ( pixelCoords, sampleIndex, dimensionIndex, frame ) => {

	// Temporal slice — true STBN temporal axis
	const slice = uint( frame ).bitAnd( uint( 63 ) ); // frame % 64

	// R2 quasi-random spatial offset for per-dimension/per-sample decorrelation
	const n = float( dimensionIndex ).add( float( sampleIndex ).mul( 7.0 ) );
	const offsetX = int( fract( n.mul( R2_A1 ).add( 0.5 ) ).mul( 128.0 ) );
	const offsetY = int( fract( n.mul( R2_A2 ).add( 0.5 ) ).mul( 128.0 ) );

	// Pixel within 128×128 tile (toroidal wrap via bitmask)
	const px = int( pixelCoords.x ).add( offsetX ).bitAnd( int( 127 ) );
	const py = int( pixelCoords.y ).add( offsetY ).bitAnd( int( 127 ) );

	// Atlas tile position from slice index
	const tileCol = int( slice ).bitAnd( int( 7 ) ); // slice % 8
	const tileRow = int( slice ).shiftRight( int( 3 ) ); // slice / 8

	return ivec2( tileCol.mul( int( 128 ) ).add( px ), tileRow.mul( int( 128 ) ).add( py ) );

};

// Sample 1D scalar STBN value in [0,1]
export const sampleSTBNScalar = ( pixelCoords, sampleIndex, dimensionIndex, frame ) => {

	const coord = computeSTBNAtlasCoord( pixelCoords, sampleIndex, dimensionIndex, frame );
	return stbnScalarTextureNode.load( coord ).x;

};

// Sample decorrelated 2D STBN pair in [0,1]²
export const sampleSTBN2D = ( pixelCoords, sampleIndex, dimensionPairIndex, frame ) => {

	const coord = computeSTBNAtlasCoord( pixelCoords, sampleIndex, dimensionPairIndex, frame );
	return stbnVec2TextureNode.load( coord ).xy;

};

// -----------------------------------------------------------------------------
// Low-discrepancy sequence generators
// -----------------------------------------------------------------------------
// Halton sequence generator with per-digit additive scrambling

export const haltonScrambled = /*@__PURE__*/ wgslFn( `
	fn haltonScrambled( index: i32, base: i32, scramble: u32 ) -> f32 {

		var result = 0.0f;
		var f = 1.0f;
		var i = index + 1;
		var s = scramble;
		var iter = 0;

		while ( i > 0 && iter < 32 ) {

			iter += 1;
			f /= f32( base );

			// Additive permutation per digit: (digit + s_k) mod base
			// Guaranteed bijection within [0, base) for any s_k
			var digit = i % base;
			digit = ( digit + i32( s % u32( base ) ) ) % base;
			result += f * f32( digit );
			i /= base;

			// Evolve scramble per digit position for position-dependent permutations
			s = s * 747796405u + 2891336453u;

		}

		return result;

	}
` );

// Owen scrambling for Sobol sequence

export const owen_scramble = /*@__PURE__*/ wgslFn( `
	fn owen_scramble( x: u32, seed: u32 ) -> u32 {

		var v = x;
		v ^= v * 0x3d20adeau;
		v += seed;
		v *= ( seed >> 16u ) | 1u;
		v ^= v >> 15u;
		v *= 0x5851f42du;
		v ^= v >> 12u;
		v *= 0x4c957f2du;
		v ^= v >> 18u;
		return v;

	}
` );

// Owen-scrambled Sobol sequence

export const owen_scrambled_sobol = /*@__PURE__*/ wgslFn( `
	fn owen_scrambled_sobol( index: u32, dimension: u32, seed: u32 ) -> f32 {

		var result = 0u;
		for ( var i = 0; i < 32; i++ ) {

			if ( ( index & ( 1u << u32( i ) ) ) != 0u ) {

				result ^= getSobolDirectionVector( i );

			}

		}

		// Mix dimension into seed for inter-dimensional decorrelation
		// (Van der Corput base is shared; Owen scrambling with distinct seeds
		// produces decorrelated sequences across dimensions)
		let dimSeed = seed ^ ( dimension * 0x9e3779b9u + 0x6a09e667u );
		result = owen_scramble( result, dimSeed );
		return f32( result ) / 4294967296.0f;

	}
`, [ getSobolDirectionVector, owen_scramble ] );

export const owen_scrambled_sobol2D = /*@__PURE__*/ wgslFn( `
	fn owen_scrambled_sobol2D( index: u32, seed: u32 ) -> vec2f {

		return vec2f(
			owen_scrambled_sobol( index, 0u, seed ),
			owen_scrambled_sobol( index, 1u, seed )
		);

	}
`, [ owen_scrambled_sobol ] );

// -----------------------------------------------------------------------------
// Multi-dimensional sampling interface
// -----------------------------------------------------------------------------
// Get N-dimensional sample (up to 4D)

export const getRandomSampleND = ( pixelCoord, sampleIndex, bounceIndex, rngState, dimensions, preferredTechnique, resolution, frame ) => {

	const technique = select( preferredTechnique.notEqual( int( - 1 ) ), preferredTechnique, samplingTechnique );
	const result = vec4( 0.0 ).toVar();

	// PCG (technique 0)
	If( technique.equal( int( 0 ) ), () => {

		// Check useFast once and branch — select() would evaluate both RandomValueFast and
		// RandomValue unconditionally, advancing state twice for each dimension.
		const useFast = dimensions.greaterThan( int( 2 ) );

		If( useFast, () => {

			result.x.assign( RandomValueFast( rngState ) );

			If( dimensions.greaterThan( int( 1 ) ), () => {

				result.y.assign( RandomValueFast( rngState ) );

			} );

			If( dimensions.greaterThan( int( 2 ) ), () => {

				result.z.assign( RandomValueFast( rngState ) );

			} );

			If( dimensions.greaterThan( int( 3 ) ), () => {

				result.w.assign( RandomValueFast( rngState ) );

			} );

		} ).Else( () => {

			result.x.assign( RandomValue( rngState ) );

			If( dimensions.greaterThan( int( 1 ) ), () => {

				result.y.assign( RandomValue( rngState ) );

			} );

			If( dimensions.greaterThan( int( 2 ) ), () => {

				result.z.assign( RandomValue( rngState ) );

			} );

			If( dimensions.greaterThan( int( 3 ) ), () => {

				result.w.assign( RandomValue( rngState ) );

			} );

		} );

	} ).ElseIf( technique.equal( int( 1 ) ), () => {

		// Halton — mix frame + bounceIndex into scramble for temporal and per-bounce decorrelation
		const pixelHash = uint( pixelCoord.x ).add( uint( pixelCoord.y ).mul( uint( resolution.x ) ) );
		const scramble = pcgHash( { state: pixelHash.bitXor( frame.mul( uint( 0x9e3779b9 ) ) ).bitXor( uint( bounceIndex ).mul( uint( 0x517cc1b7 ) ) ) } ).toVar();

		result.x.assign( haltonScrambled( { index: sampleIndex, base: int( 2 ), scramble } ) );

		If( dimensions.greaterThan( int( 1 ) ), () => {

			result.y.assign( haltonScrambled( { index: sampleIndex, base: int( 3 ), scramble } ) );

		} );

		If( dimensions.greaterThan( int( 2 ) ), () => {

			result.z.assign( haltonScrambled( { index: sampleIndex, base: int( 5 ), scramble } ) );

		} );

		If( dimensions.greaterThan( int( 3 ) ), () => {

			result.w.assign( haltonScrambled( { index: sampleIndex, base: int( 7 ), scramble } ) );

		} );

	} ).ElseIf( technique.equal( int( 2 ) ), () => {

		// Sobol — mix frame + bounceIndex into seed for temporal and per-bounce decorrelation
		const pixelHash = uint( pixelCoord.x ).add( uint( pixelCoord.y ).mul( uint( resolution.x ) ) );
		const seed = pcgHash( { state: pixelHash.bitXor( frame.mul( uint( 0x9e3779b9 ) ) ).bitXor( uint( bounceIndex ).mul( uint( 0x517cc1b7 ) ) ) } ).toVar();

		result.x.assign( owen_scrambled_sobol( { index: uint( sampleIndex ), dimension: uint( 0 ), seed } ) );

		If( dimensions.greaterThan( int( 1 ) ), () => {

			result.y.assign( owen_scrambled_sobol( { index: uint( sampleIndex ), dimension: uint( 1 ), seed } ) );

		} );

		If( dimensions.greaterThan( int( 2 ) ), () => {

			result.z.assign( owen_scrambled_sobol( { index: uint( sampleIndex ), dimension: uint( 2 ), seed } ) );

		} );

		If( dimensions.greaterThan( int( 3 ) ), () => {

			result.w.assign( owen_scrambled_sobol( { index: uint( sampleIndex ), dimension: uint( 3 ), seed } ) );

		} );

	} ).Else( () => {

		// STBN — Spatiotemporal Blue Noise (technique 3)
		// Each bounce uses a block of 4 dimension indices for decorrelation
		const dimBase = bounceIndex.mul( int( 4 ) );

		If( dimensions.lessThanEqual( int( 2 ) ), () => {

			const _sample = sampleSTBN2D( pixelCoord, sampleIndex, dimBase, frame );
			result.x.assign( _sample.x );
			result.y.assign( _sample.y );

		} ).Else( () => {

			const _sample1 = sampleSTBN2D( pixelCoord, sampleIndex, dimBase, frame );
			const _sample2 = sampleSTBN2D( pixelCoord, sampleIndex, dimBase.add( int( 1 ) ), frame );
			result.assign( vec4( _sample1, _sample2 ) );

		} );

	} );

	return result;

};

// -----------------------------------------------------------------------------
// Main sampling interface functions
// -----------------------------------------------------------------------------
// Get random sample based on preferred technique (2D)

export const getRandomSample = ( pixelCoord, sampleIndex, bounceIndex, rngState, preferredTechnique, resolution, frame ) => {

	const sample4D = getRandomSampleND( pixelCoord, sampleIndex, bounceIndex, rngState, int( 2 ), preferredTechnique, resolution, frame );

	return sample4D.xy;

};

// Get stratified sample with proper blue noise support

export const getStratifiedSample = ( pixelCoord, rayIndex, totalRays, rngState, resolution, frame ) => {

	// result variable avoids early-return ReturnNode escaping into outer Fn scope
	const result = vec2( 0.0 ).toVar();

	If( totalRays.lessThanEqual( int( 1 ) ), () => {

		result.assign( getRandomSample( pixelCoord, rayIndex, int( 0 ), rngState, int( - 1 ), resolution, frame ) );

	} ).Else( () => {

		// Calculate strata dimensions

		const strataX = int( sqrt( float( totalRays ) ) );
		const strataY = totalRays.add( strataX ).sub( 1 ).div( strataX );
		const strataIdx = mod( rayIndex, strataX.mul( strataY ) );
		const sx = mod( strataIdx, strataX );
		const sy = strataIdx.div( strataX );

		// Base stratified position

		const strataPos = vec2( float( sx ), float( sy ) ).div( vec2( float( strataX ), float( strataY ) ) );

		// Jitter via STBN or fast RNG fallback

		const jitter = vec2( 0.0 ).toVar();

		If( samplingTechnique.greaterThanEqual( int( 3 ) ), () => {

			// STBN — true spatiotemporal blue noise jitter
			jitter.assign( sampleSTBN2D( pixelCoord, rayIndex, int( 0 ), frame ) );

		} ).Else( () => {

			// Fast RNG with subtle STBN influence for better convergence
			const j1 = RandomValueFast( rngState ).toVar();
			const j2 = RandomValueFast( rngState ).toVar();
			jitter.assign( vec2( j1, j2 ) );

			If( totalRays.greaterThan( int( 4 ) ), () => {

				const stbnInfluence = sampleSTBN2D( pixelCoord, rayIndex, int( 0 ), frame ).mul( 0.1 );
				jitter.assign( mix( jitter, stbnInfluence, 0.2 ) );

			} );

		} );

		jitter.divAssign( vec2( float( strataX ), float( strataY ) ) );

		result.assign( strataPos.add( jitter ) );

	} );

	return result;

};

// Get decorrelated seed with better mixing

export const getDecorrelatedSeed = /*@__PURE__*/ wgslFn( `
	fn getDecorrelatedSeed( pixelCoord: vec2f, rayIndex: i32, frame: u32 ) -> u32 {

		// Use multiple primes for better decorrelation
		let pixelSeed = u32( pixelCoord.x ) * 2654435761u + u32( pixelCoord.y ) * 3266489917u;
		let raySeed = u32( rayIndex ) * 668265263u;
		let frameSeed = frame * 374761393u;

		// Multiple rounds of hashing for better quality
		var seed = wang_hash( pixelSeed );
		seed = pcgHash( seed ^ raySeed );
		seed = wang_hash( seed + frameSeed );
		return seed;

	}
`, [ wang_hash, pcgHash ] );

